src/Psalm/Internal/Type/SimpleAssertionReconciler.php

```diff
@@ -2021,20 +2025,53 @@ private static function reconcileList(
         $did_remove_type = false;
 
         foreach ($existing_var_atomic_types as $type) {
-            if ($type instanceof TList
-                || ($type instanceof TKeyedArray && $type->is_list)
-            ) {
-                if ($is_non_empty && $type instanceof TList && !$type instanceof TNonEmptyList) {
+            if ($type instanceof TList) {
+                if ($is_non_empty && !$type instanceof TNonEmptyList) {
                     $array_types[] = new TNonEmptyList($type->type_param);
                     $did_remove_type = true;
                 } else {
                     $array_types[] = $type;
                 }
-            } elseif ($type instanceof TArray || $type instanceof TKeyedArray) {
-                if ($type instanceof TKeyedArray) {
-                    $type = $type->getGenericArrayType();
-                }
+            } elseif ($type instanceof TKeyedArray) {
+                if ($type->is_list) {
+                    $array_types[] = $type;
+                } else {
+                    $did_remove_type = true;
+                    $type = clone $type;
+                    $min_unset_list_key = 0; // Minimum list key not explicitly set
+                    ksort($type->properties);
+                    foreach ($type->properties as $prop_key => $prop_value) {
```

Comment:
Oh yeah, I guess that's a better way of fixing that than just adding the count property